aggregator/list.go

```diff
@@ -182,34 +180,37 @@ func (l *metricList) PushBack(value interface{}) (*list.Element, error) {
 // Close closes the list.
 func (l *metricList) Close() {
 	l.Lock()
-	defer l.Unlock()
-
 	if l.closed {
+		l.Unlock()
 		return
 	}
 	l.closed = true
-}
-
-func (l *metricList) Flush() {
-	flushStart := l.nowFn()
+	l.Unlock()
 
-	// NB(xichen): it is important to determine ticking start time within the time lock
-	// because this ensures all the actions before `start` have completed if those actions
-	// are protected by the same read lock.
-	l.timeLock.Lock()
-	consumeStart := l.nowFn()
-	resolution := l.resolution
-	l.timeLock.Unlock()
-	alignedStartNanos := consumeStart.Truncate(resolution).UnixNano()
-	l.flushBefore(alignedStartNanos, consumeType)
+	// Unregisters the list outside the list lock to avoid holding the list lock
+	// while attempting to acquire the flush manager lock potentially causing a
+	// deadlock.
+	if err := l.flushMgr.Unregister(l); err != nil {
+		l.log.Errorf("error unregistering list: %v", err)
+	}
+}
 
-	flushDuration := l.nowFn().Sub(flushStart)
-	l.metrics.flushDuration.Record(flushDuration)
+func (l *metricList) Flush(cutoverNanos, cutoffNanos int64) {
+	start := l.nowFn()
+	alignedNowNanos := l.alignedNowNanos()
+	if alignedNowNanos > cutoverNanos && alignedNowNanos <= cutoffNanos {
```

Comment:
so we are flushing (cutover, cutoff]? should it be [cutover, cutoff]?

Reply:
Yeah so the rationale here is that if `alignedNowNanos == cutoverNanos`, we should discard everything before cutoverNanos because the metrics before `cutoverNanos` have not been cut over yet so should be discarded, and if `alignedNowNanos == cutoffNanos`, we should still flush them because metrics before `cutoffNanos` have not been cut off yet so should still be flushed.

Reply:
I see, would be good if you could put add some comments for this.

Reply:
Sure thing.